### PR TITLE
Add feature to convert between regular and verbatim string literals.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/42061

Looks like this:

image

Works in reverse:

image

Todo:

- [x] Work on interpolated strings.
- [x] tests

### CODE REVIEW

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertBetweenRegularAndVerbatimString
+{
+    internal abstract class AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider<
```
Comment (tmat) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]: It seems like an overkill to include the name of the refactoring in both the containing namespace and the typename.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]: > It seems like an overkill to include the name of the refactoring in both the containing namespace and the typename.

This is how we name all our refactorings/analyzers/fixers.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]: > I'd expect us to have all these under a common namespace 

We don't do that.  We might make that happen when/if we move this to lower layers or into a dll dedicated to refactorings.

**Comment by CyrusNajmabadi** [comment]
Tests written.  This is ready for review.

File: src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringTests.cs [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.ConvertBetweenRegularAndVerbatimString;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertForEachToFor
+{
+    public class ConvertBetweenRegularAndVerbatimInterpolatedStringTests : AbstractCSharpCodeActionTest
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringTests.cs]: @sharwell I recall you mentioned adding support for writing code refactoring unit tests on top of Microsoft.CodeAnalysis.Testing. Is that available in latest package? Should we encourage writing new unit tests on top of that API?

File: src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringTests.cs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.ConvertBetweenRegularAndVerbatimString;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertForEachToFor
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringTests.cs]:
```suggestion
namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertBetweenRegularAndVerbatimString
```

File: src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringTests.cs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.ConvertBetweenRegularAndVerbatimString;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertForEachToFor
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringTests.cs]:
```suggestion
namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertBetweenRegularAndVerbatimString
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            {
+                // Offer to convert to a verbatim string if the normal string contains simple
+                // escapes that can be directly embedded in the verbatim string.
+                context.RegisterRefactoring(new MyCodeAction(
+                    CSharpFeaturesResources.Convert_to_verbatim_string,
+                    c => ConvertToVerbatimStringAsync(document, literalExpression, c)));
+            }
+        }
+
+        private async Task<Document> ConvertAsync(
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]:
```suggestion
        private static async Task<Document> ConvertAsync(
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                // This was checked above.
+                Debug.Assert(!chars.IsDefault);
+                if (ContainsSimpleEscape(chars))
+                    return true;
+            }
+
+            return false;
+        }
+
+        private bool ContainsSimpleEscape(VirtualCharSequence chars)
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]:
```suggestion
        private static bool ContainsSimpleEscape(VirtualCharSequence chars)
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                    sb.Append(ch);
+
+                    // if it's an interpolation, we need to double-up open/close braces.
+                    if (isInterpolation && IsOpenOrCloseBrace(ch))
+                        sb.Append(ch);
+                }
+            }
+        }
+
+        private bool ContainsSimpleEscape(
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]:
```suggestion
        private static bool ContainsSimpleEscape(
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertBetweenRegularAndVerbatimString
+{
+    internal abstract class AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider<
+        TStringExpressionSyntax>
+        : CodeRefactoringProvider
+        where TStringExpressionSyntax : ExpressionSyntax
+    {
+        protected const char OpenBrace = '{';
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]:
```suggestion
        private const char OpenBrace = '{';
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        protected const char CloseBrace = '}';
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]:
```suggestion
        private const char CloseBrace = '}';
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        protected override void AddSubStringTokens(LiteralExpressionSyntax literalExpression, ArrayBuilder<SyntaxToken> subStringTokens)
+            => subStringTokens.Add(literalExpression.Token);
+
+        protected override bool IsVerbatim(LiteralExpressionSyntax literalExpression)
+            => CSharpSyntaxFacts.Instance.IsVerbatimStringLiteral(literalExpression.Token);
+
+        protected override LiteralExpressionSyntax CreateVerbatimStringExpression(IVirtualCharService charService, StringBuilder sb, LiteralExpressionSyntax stringExpression)
+        {
+            sb.Append('@');
+            sb.Append(DoubleQuote);
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]: Nit: can this be implemented as:

```csharp
   protected override LiteralExpressionSyntax CreateVerbatimStringExpression(IVirtualCharService charService, StringBuilder sb, LiteralExpressionSyntax stringExpression)
   {
      sb.Append('@');
      return CreateRegularStringExpression(charService, sb, stringExpression);
   }
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringCodeRefactoringProvider.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                kind: newStartKind,
+                trailing: startToken.TrailingTrivia);
+
+            return stringExpression.Update(
+                newStartToken,
+                SyntaxFactory.List(newContents),
+                stringExpression.StringEndToken);
+        }
+
+        private SyntaxToken CreateTextToken(SyntaxToken textToken, StringBuilder sb)
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringCodeRefactoringProvider.cs]:
```suggestion
        private static SyntaxToken CreateTextToken(SyntaxToken textToken, StringBuilder sb)
```

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringCodeRefactoringProvider.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                if (content is InterpolatedStringTextSyntax textSyntax)
+                    subStringTokens.Add(textSyntax.TextToken);
+            }
+        }
+
+        protected override bool IsVerbatim(InterpolatedStringExpressionSyntax literalExpression)
+            => literalExpression.StringStartToken.Kind() == SyntaxKind.InterpolatedVerbatimStringStartToken;
+
+        private InterpolatedStringExpressionSyntax Convert(
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringCodeRefactoringProvider.cs]:
```suggestion
        private static InterpolatedStringExpressionSyntax Convert(
```

File: src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringTests.cs]: done.

File: src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringTests.cs]: done.

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/AbstractConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]: done.

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimStringCodeRefactoringProvider.cs]: not going to do this.  i think the current approach is clearer. thanks :)

File: src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringCodeRefactoringProvider.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertBetweenRegularAndVerbatimString/ConvertBetweenRegularAndVerbatimInterpolatedStringCodeRefactoringProvider.cs]: done.

**Pull request merged**